docs/design_doc.md

```diff
@@ -0,0 +1,110 @@
+
+    * When the mirror is redeployed, the cache will have to be rebuilt once only, instead of once for *each* replica.
+
+* Use an in-memory key-value store as the shared cache server.
+
+    * The etags returned with each conditional request to the Github API are unique. Therefore etags can be used as a key to store and retreive the results of each API request. 
+
+* The key-value store should meet the following requirements.
+
+    * Highly available, to prevent bottlenecks and minimize latency.
+    * Persistent, to prevent loss of all cached information in the case of a pod restart.
+    * Weak consistency since we can tolerate stale reads
+
+### Implementation details
+
+* The shared cache will be implemented using a Redis cluster. 
+
+    * Since the goal is to minimize response time, a in-memory key-value is needed. There are several key-value stores that meet the critera, notably [Redis](https://redis.io/) or [Memcached](https://memcached.org/).
+
+    * Memcached has a limit on the size of the object being stored (< 1 MB), however, most of the response objects returned by the API will be small and therefore Memcached or Redis will both work in this regard.
+
+    * The main advantage that Redis has over Memcached is with regards to availability and persistence. [Memcached does not offer replication or persistence](https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/SelectEngine.html), so if the server fails or is restarted, the cache will be lost.
+
+    * Redis offers [persistence](https://redis.io/topics/persistence) using <a name="rdb-files"></a>RDB files which are snapshots of the data written to disk at specifiable intervals. Redis provides more durable persistence using AOF logs but that is not required in this case.
+
+    * Redis can be configured to be highly available. A Redis master can be replicated using two worker nodes, but can replace it in the case of a failover. A Redis sentinel is used to select the replacement node when a master node fails.
+
+    * Redis does not guarantee [strong consistency](https://redis.io/topics/cluster-tutorial#redis-cluster-consistency-guarantees) which is acceptable since any stale cache reads will always be compared against the response from the conditional request, resulting - in the worst case - in an additional request to the Github API as well as unnecessarily replacing the cache entry.
+
+    * Other features of Redis include sharding the cache if it gets too big, as well as the ability to select from a number of cache eviction policies.
+
+* Measuring latency
+
+    * A Prometheus metrics endpoint exists that is used to graph the average latency of the mirror for the 90, 95, 99 percentile. Another metric can be added to show the additional latency created by introducing the Redis cache server.
+
+### Architecture diagram
+
+![](./images/arch.png)
+
+[Image Source](https://docs.google.com/document/d/1WrZFnto4Rin0hYfJwNYIKVO-5HLzOC6cFM7NnJ-Wuak/edit?usp=sharing)
```

Comment:
Can you export to SVG then commit the SVG file? Non-RH people can't access this doc.

Reply:
ok that's a good idea